libpoolprotocols/PoolManager.cpp

```diff
@@ -319,7 +316,7 @@ void PoolManager::setActiveConnection(unsigned int idx)
     }
 }
 
-const URI *PoolManager::getActiveConnection()
+const URI* PoolManager::getActiveConnection()
 {
     if (m_connections.size() > m_activeConnectionIdx)
         return &m_connections.at(m_activeConnectionIdx);
```

Comment:
More like C++ would be:

`return &m_connections[m_activeConnectionIdx];`

Reply:
Or, simply:

`return m_connections.at + m_activeConnectionIdx;`

Reply:
I'm always puzzled about direct random access on vectors. Any index out of bounds does not throw while .at() method does.

Personally I'd simplify the with the single line

`return &m_connections.at(m_activeConnectionIdx);`

as m_activeConnectionIdx is checked to be inbounds every time is set.

Reply:
That is precisely why I dislike the .at() notation. The out-of-bounds case is already taken care of by the preceding if, and I prefer the implied understanding of pointer arithmetic of using the perfectly equivalent:

`return m_connections + m_activeConnectionIdx;`

Reply:
Strike that... It's a vector, not an array. In which case I'd use:

`return &m_connections[m_activeConnectionIdx];`

I don't feel strongly about this. In my view, you either check for out-of-bound conditions before you use the index, or establish a proper try/catch construct and use the .at() accessor. Not both!

Reply:
My personal favorite was `return &m_connections[m_activeConnectionIdx];` .
But I already implemented that a time ago while @AndreaLanfranchi  were searching bugs in the auto/connection. I think he replaced while searching the bug nearly all indexed array accesses (m_connections[]) with m_connections.at().
Only one m_connections[] access is currently in the PoolManager.cpp
I don't wanted to break the style so I also used m_connections.at().

Would it be possible that thread2 eliminates the element (after thread1 has checked the size ) and the then thread1 accesses an invalid element (eg if it was the last)?

Reply:
m_activeConnectionIdx can be set by 2 threads (poolmgr and API) as well as the size of the vector although it's very rare for poolManager and API calls are supposed to be serialized (unless more than 1 API connection tries to issue commands ... which I expect unlikely).
For _extreme_ safety I'd go for a try/catch using .at() but in that case ... how do you think to handle the exception ? I mean ... which element would you return ?

Probably better to let getActiveConnection() throw and handle exception at higher level.
My 2 cents.